src/hotspot/share/gc/g1/g1ConcurrentMark.hpp

```diff
@@ -537,6 +538,9 @@ class G1ConcurrentMark : public CHeapObj<mtGC> {
 
   // Region statistics gathered during marking.
   G1RegionMarkStats* _region_mark_stats;
+  // Top pointer for each region at the start of marking. Must be valid for all committed
+  // regions.
+  HeapWord* volatile* _top_at_mark_starts;
```

Comment:
I wonder if one can use `HeapWord* volatile _top_at_mark_starts[];` to emphasize that it's an array.

Reply:
No, it would be interpreted as flexible array, which would need to be the last member of the class.